src/main/java/spoon/reflect/factory/ModuleFactory.java

```diff
@@ -0,0 +1,163 @@
+/**
+ * Copyright (C) 2006-2017 INRIA and contributors
+ * Spoon - http://spoon.gforge.inria.fr/
+ *
+ * This software is governed by the CeCILL-C License under French law and
+ * abiding by the rules of distribution of free software. You can use, modify
+ * and/or redistribute the software under the terms of the CeCILL-C license as
+ * circulated by CEA, CNRS and INRIA at http://www.cecill.info.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the CeCILL-C License for more details.
+ *
+ * The fact that you are presently reading this means that you have had
+ * knowledge of the CeCILL-C license and that you accept its terms.
+ */
+package spoon.reflect.factory;
+
+import spoon.reflect.CtModelImpl;
+import spoon.reflect.declaration.CtElement;
+import spoon.reflect.declaration.CtModule;
+import spoon.reflect.declaration.CtPackageExport;
+import spoon.reflect.declaration.CtProvidedService;
+import spoon.reflect.declaration.CtNamedElement;
+import spoon.reflect.declaration.CtUsedService;
+import spoon.reflect.declaration.ParentNotInitializedException;
+import spoon.reflect.reference.CtModuleReference;
+import spoon.reflect.declaration.CtModuleRequirement;
+import spoon.reflect.reference.CtPackageReference;
+import spoon.reflect.reference.CtTypeReference;
+import spoon.reflect.visitor.CtVisitor;
+import spoon.support.reflect.declaration.CtElementImpl;
+import spoon.support.reflect.declaration.CtModuleImpl;
+
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+public class ModuleFactory extends SubFactory implements Serializable {
+
+	public static class CtUnnamedModule extends CtModuleImpl {
+		final Map<String, CtModule> allModules = new HashMap<>();
+
+		{
+			this.setSimpleName(CtModule.TOP_LEVEL_MODULE_NAME);
+			this.setParent(new CtElementImpl() {
+				@Override
+				public void accept(CtVisitor visitor) {
+
+				}
+
+				@Override
+				public CtElement getParent() throws ParentNotInitializedException {
+					return null;
+				}
+
+				@Override
+				public Factory getFactory() {
+					return CtUnnamedModule.this.getFactory();
+				}
+			});
+
+			this.setRootPackage(new CtModelImpl.CtRootPackage());
+			this.addModule(this);
+		}
+
+		public boolean addModule(CtModule module) {
+			if (allModules.containsKey(module.getSimpleName())) {
+				return false;
+			}
+			allModules.put(module.getSimpleName(), module);
```

Comment:
I see one more problem. Later call of CtModule.setSimpleName(...) will cause inconsistency in modules, because the module is registered under old name. 
May be we should store modules as List<CtModule> and `getModule` should always search that list by name.
Then you can add the module into that list sooner directly in `factory.Core().createModule()` call - at time when it has no name. WDYT?

Reply:
> I see one more problem. Later call of CtModule.setSimpleName(...) will cause inconsistency in modules, because the module is registered under old name.

Agree that's a problem. I'll try to produce a test to show the bug, and fix it the way you propose.